### PR TITLE
Add method for SDWebImageDownloaderOperation subclass to decrypt or decompress image data

### PR DESCRIPTION
**Pull request by ghughes**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [ ] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / reffers to the following issues: see below

### Pull Request Description

This adds a `prepareImageData` method to SDWebImageDownloaderOperation that is intended for subclasses to override if they need to do something special with the image data after it's downloaded but before it's cached or converted to a UIImage. I'm using it in my project for decryption but it could also be useful for custom decompression. Without this change, it's hard to implement these kinds of behaviors in a SDWebImageDownloaderOperation subclass.

### CODE REVIEW

**Comment by bpoplauschi** [comment]
@ghughes this looks like a good improvement to our code. I will look for a future release to add this.

**Comment by ghughes** [comment]
Looks like the test failure was environmental, FYI.

File: SDWebImage/SDWebImageDownloaderOperation.h [hunk 1]
```diff
@@ -97,7 +97,7 @@ extern NSString * _Nonnull const SDWebImageDownloadFinishNotification;
                                 options:(SDWebImageDownloaderOptions)options NS_DESIGNATED_INITIALIZER;
 
 /**
- *  Adds handlers for progress and completion. Returns a tokent that can be passed to -cancel: to cancel this set of
```
Comment (VinceXu) [inline SDWebImage/SDWebImageDownloaderOperation.h]: ...

Comment (ghughes) [inline SDWebImage/SDWebImageDownloaderOperation.h]: I fixed a typo: "tokent"

**Comment by dreampiggy** [comment]
Maybe we can go further ? I recently refactory the entire decoding process in SDWebImage. I provide an Protocol to let user to provide any decode method. You can set it on `SDImageCache` or `SDWebImageDownloader`.

For your usage, you can write your decoder like

```objectivec
@interface DecryptCoder : NSObject <SDWebImageCoder>

@end

@implementation DecryptCoder

- (UIImage *)decompressedImageWithImage:(UIImage *)image data:(NSData *__autoreleasing  _Nullable *)data format:(SDImageFormat)format shouldScaleDown:(BOOL)shouldScaleDown {
    *data = [[SomeEntryptData alloc] init];
    UIImage *newImage = someDecompressMethod(image);
    return newImage;
}

@end
```

```objectivec
[[SDImageCache sharedCache] setImageCoder:[[DecryptCoder alloc] init]];
[[SDWebImageDownloader sharedDownloader] setImageCoder:[[DecryptCoder alloc] init]];
```

You can view [here](https://github.com/rs/SDWebImage/pull/1987) and check [these](https://github.com/dreampiggy/SDWebImage/blob/feature_refactor_decoding_with_plugin/SDWebImage/SDWebImageDecoder.m) to discuss.

**Comment by bpoplauschi** [comment]
Update to the latest comment: you can just add that coder to the list of coders. No need to merge this PR.
```[[SDWebImageCodersManager sharedInstance] addCoder: [DecryptCoder new]];```

**Pull request closed without merging**